sdks/python/apache_beam/transforms/environments.py

```diff
@@ -0,0 +1,331 @@
+
+    def decorator(env_cls):
+
+      def to_runner_api_parameter(self, context):
+        return (urn, self.payload)
+
+      setattr(env_cls, 'to_runner_api_parameter', to_runner_api_parameter)
+
+      @Environment.register_urn(urn, param_type)
+      def from_runner_api_parameter(payload, unused_context):
+        if param_type is bytes:
+          payload = payload.decode('utf-8')
+        if payload:
+          return env_cls(payload=payload)
+        else:
+          return env_cls()
+
+      setattr(env_cls, 'from_runner_api_parameter', from_runner_api_parameter)
+
+      cls._urn_to_env_cls[urn] = env_cls
+      return env_cls
+
+    return decorator
+
+  @classmethod
+  def get_env_cls_from_urn(cls, urn):
+    return cls._urn_to_env_cls[urn]
+
+  def to_runner_api(self, context):
+    urn, typed_param = self.to_runner_api_parameter(context)
+    return beam_runner_api_pb2.Environment(
+        urn=urn,
+        payload=typed_param.SerializeToString()
+        if isinstance(typed_param, message.Message)
+        else typed_param.encode('utf-8') if isinstance(typed_param, str)
+        else typed_param
+    )
+
+  @classmethod
+  def from_runner_api(cls, proto, context):
+    if proto is None or not proto.urn:
+      return None
+    parameter_type, constructor = cls._known_urns[proto.urn]
+
+    try:
+      return constructor(
+          proto_utils.parse_Bytes(proto.payload, parameter_type),
+          context)
+    except Exception:
+      if context.allow_proto_holders:
+        return RunnerAPIEnvironmentHolder(proto)
+      raise
+
+  @classmethod
+  def from_options(cls, options):
+    """Creates an Environment object from PipelineOptions.
+
+    Args:
+      options: The PipelineOptions object.
+    """
+    raise NotImplementedError
+
+
+class DockerEnvironment(Environment):
+
+  def __init__(self, container_image=None):
+    from apache_beam.runners.portability.portable_runner import PortableRunner
+
+    if container_image:
+      self.container_image = container_image
+    else:
+      self.container_image = PortableRunner.default_docker_image()
+
+  def __eq__(self, other):
+    return self.__class__ == other.__class__ \
+           and self.container_image == other.container_image
+
+  def __ne__(self, other):
+    # TODO(BEAM-5949): Needed for Python 2 compatibility.
+    return not self == other
+
+  def __repr__(self):
+    return 'DockerEnvironment(container_image=%s)' % self.container_image
+
+  def to_runner_api_parameter(self, context):
+    return (common_urns.environments.DOCKER.urn,
+            beam_runner_api_pb2.DockerPayload(
+                container_image=self.container_image))
+
+  @staticmethod
+  def from_runner_api_parameter(payload, context):
+    return DockerEnvironment(container_image=payload.container_image)
+
+  @classmethod
+  def from_options(cls, options):
+    return cls(container_image=options.environment_config)
+
+
+Environment.register_urn(common_urns.environments.DOCKER.urn,
```

Comment:
I believe we should still be able to use this as a decorator, as with `PTransform.register_urn`